crates/driver_net/src/ixgbe.rs

```diff
@@ -0,0 +1,157 @@
+use core::{mem::ManuallyDrop, ptr::NonNull};
+
+use alloc::{collections::VecDeque, sync::Arc};
+use driver_common::{BaseDriverOps, DevError, DevResult, DeviceType};
+use ixgbe_driver::{IxgbeDevice, IxgbeError, IxgbeNetBuf, MemPool, NicDevice};
+pub use ixgbe_driver::{IxgbeHal, PhysAddr, INTEL_82599, INTEL_VEND};
+
+use crate::{EthernetAddress, NetBufPtr, NetDriverOps};
+
+extern crate alloc;
+
+const RECV_BATCH_SIZE: usize = 64;
+const RX_BUFFER_SIZE: usize = 1024;
+const MEM_POOL: usize = 4096;
+const MEM_POOL_ENTRY_SIZE: usize = 2048;
+
+/// The ixgbe NIC device driver.
+///
+/// `QS` is the ixgbe queue size.
+pub struct IxgbeNic<H: IxgbeHal, const QS: u16> {
+    inner: IxgbeDevice<H>,
+    mem_pool: Arc<MemPool>,
+    rx_buffer_queue: VecDeque<NetBufPtr>,
+}
+
+unsafe impl<H: IxgbeHal, const QS: u16> Sync for IxgbeNic<H, QS> {}
+unsafe impl<H: IxgbeHal, const QS: u16> Send for IxgbeNic<H, QS> {}
+
+impl<H: IxgbeHal, const QS: u16> IxgbeNic<H, QS> {
+    /// Creates a net ixgbe NIC instance and initialize, or returns a error if
+    /// any step fails.
+    pub fn init(base: usize, len: usize) -> DevResult<Self> {
+        let mem_pool = MemPool::allocate::<H>(MEM_POOL, MEM_POOL_ENTRY_SIZE)
+            .map_err(|_| DevError::NoMemory)?;
+        let inner = IxgbeDevice::<H>::init(base, len, QS, QS, &mem_pool).map_err(|err| {
+            error!("Failed to initialize ixgbe device: {:?}", err);
+            DevError::BadState
+        })?;
+
+        let rx_buffer_queue = VecDeque::with_capacity(RX_BUFFER_SIZE);
+        Ok(Self {
+            inner,
+            mem_pool,
+            rx_buffer_queue,
+        })
+    }
+}
+
+impl<H: IxgbeHal, const QS: u16> BaseDriverOps for IxgbeNic<H, QS> {
+    fn device_name(&self) -> &str {
+        self.inner.get_driver_name()
+    }
+
+    fn device_type(&self) -> DeviceType {
+        DeviceType::Net
+    }
+}
+
+impl<H: IxgbeHal, const QS: u16> NetDriverOps for IxgbeNic<H, QS> {
+    fn mac_address(&self) -> EthernetAddress {
+        EthernetAddress(self.inner.get_mac_addr())
+    }
+
+    fn rx_queue_size(&self) -> usize {
+        self.inner.num_rx_queues() as usize
+    }
+
+    fn tx_queue_size(&self) -> usize {
+        self.inner.num_tx_queues() as usize
+    }
+
+    fn can_receive(&self) -> bool {
+        self.inner.can_receive(0).unwrap() || !self.rx_buffer_queue.is_empty()
+    }
+
+    fn can_transmit(&self) -> bool {
+        // Default implementation is return true forever.
+        self.inner.can_send(0).unwrap()
+    }
+
+    fn recycle_rx_buffer(&mut self, rx_buf: NetBufPtr) -> DevResult {
+        let rx_buf =
```

Comment:
Add a method to convert `NetBufPtr` to `IxgbeNetBuf`

Reply:
Since `IxgbeNetBuf` is defined in the `ixgbe-driver` crate and `NetBufPtr` is a generic data structure, implementing it as member methods of both seems inappropriate. Perhaps it could be implemented as external non-member functions?

Reply:
It seems that implementing `From` and `Into` for `NetBufPtr` is possible.